sstable/reader_virtual.go

```diff
@@ -204,10 +248,264 @@ func (v *virtualState) constrainBounds(
+	if !bytes.HasPrefix(key, p.dst) {
+		p.err = errInputPrefixMismatch
+		return key
+	}
+	p.arg2 = append(p.arg2[:len(p.src)], key[len(p.dst):]...)
+	return p.arg2
+}
+
+func (p *prefixReplacingIterator) rewriteResult(
+	k *InternalKey, v base.LazyValue,
+) (*InternalKey, base.LazyValue) {
+	if k == nil {
+		return k, v
+	}
+	if !bytes.HasPrefix(k.UserKey, p.src) {
+		panic(errOutputPrefixMismatch)
+	}
+	p.res.Trailer = k.Trailer
+	p.res.UserKey = append(p.res.UserKey[:len(p.dst)], k.UserKey[len(p.src):]...)
+	return &p.res, v
+}
+
+// SeekGE implements the Iterator interface.
+func (p *prefixReplacingIterator) SeekGE(
+	key []byte, flags base.SeekGEFlags,
+) (*InternalKey, base.LazyValue) {
+	cmp := p.cmp(key, p.dst)
+	if cmp < 0 {
+		return p.rewriteResult(p.i.SeekGE(p.src, flags))
+	}
+	return p.rewriteResult(p.i.SeekGE(p.rewriteArg(key), flags))
+}
+
+// SeekPrefixGE implements the Iterator interface.
+func (p *prefixReplacingIterator) SeekPrefixGE(
+	prefix, key []byte, flags base.SeekGEFlags,
+) (*InternalKey, base.LazyValue) {
+	cmp := p.cmp(key, p.dst)
+	if cmp < 0 {
+		// Prefix mismatch by definition.
+		return nil, base.LazyValue{}
+	}
+	return p.rewriteResult(p.i.SeekPrefixGE(p.rewriteArg2(prefix), p.rewriteArg(key), flags))
+}
+
+// SeekLT implements the Iterator interface.
+func (p *prefixReplacingIterator) SeekLT(
+	key []byte, flags base.SeekLTFlags,
+) (*InternalKey, base.LazyValue) {
+	cmp := p.cmp(key, p.dst)
+	if cmp < 0 {
+		// Exhaust the iterator by Prev()ing before the First key.
+		p.i.First()
+		return p.rewriteResult(p.i.Prev())
+	}
+	return p.rewriteResult(p.i.SeekLT(p.rewriteArg(key), flags))
+}
+
+// First implements the Iterator interface.
+func (p *prefixReplacingIterator) First() (*InternalKey, base.LazyValue) {
+	return p.rewriteResult(p.i.First())
+}
+
+// Last implements the Iterator interface.
+func (p *prefixReplacingIterator) Last() (*InternalKey, base.LazyValue) {
+	return p.rewriteResult(p.i.Last())
+}
+
+// Next implements the Iterator interface.
+func (p *prefixReplacingIterator) Next() (*InternalKey, base.LazyValue) {
+	return p.rewriteResult(p.i.Next())
+}
+
+// NextPrefix implements the Iterator interface.
+func (p *prefixReplacingIterator) NextPrefix(succKey []byte) (*InternalKey, base.LazyValue) {
+	return p.rewriteResult(p.i.NextPrefix(p.rewriteArg(succKey)))
+}
+
+// Prev implements the Iterator interface.
+func (p *prefixReplacingIterator) Prev() (*InternalKey, base.LazyValue) {
+	return p.rewriteResult(p.i.Prev())
+}
+
+// Error implements the Iterator interface.
+func (p *prefixReplacingIterator) Error() error {
+	if p.err != nil {
+		return p.err
+	}
+	return p.i.Error()
+}
+
+// Close implements the Iterator interface.
+func (p *prefixReplacingIterator) Close() error {
+	return p.i.Close()
+}
+
+// SetBounds implements the Iterator interface.
+func (p *prefixReplacingIterator) SetBounds(lower, upper []byte) {
+	p.i.SetBounds(p.rewriteArg(lower), p.rewriteArg2(upper))
```

Comment:
This is redundant / slightly buggy as the underlying iterator will do the same rewrite as part of `virtualState.constrainBounds`.

Reply:
Something feels a bit off here? PrefixReplacingIterator claims to implement Iterator, which includes SetBounds. The contract is that takes "new" prefix keys and bounds results to be within them. 

The underlying iterator is over "old" keys, and has its contract for SetBounds , again in terms of old keys. They underlying iterator doesn't know it is being wrapped in a prefix iterator. 

So why would I pass "new" key bounds to the "old" key iterator?

Reply:
You're right about the iterator _interface_. The issue is that the sstable iterators (both singleLevel and twoLevel) have a notion of what the VSST file bounds are (let's say a VSST has bounds `[c-e)`), and they get a set of bounds from the interface (let's say `[d-g)`), and the sstable iterator will call `constrainBounds` which will go "oh I need to impose `[d-e)` on myself because anything outside of that in this sstable could be garbage".

The VSST bounds in the `FileMetadata`, which is what `constrainBounds` uses, are in the post-transformation keyspace. The bounds coming from the iterator, with your change, were sometimes in the post-transformation keyspace (if they were passed in as a creation-time IterOption), or pre-transformation (if they got rewriteArg'd in SetBounds). In the latter case, constrainBounds would go "oh, the user wants `[de-di)`, I have `[ef-ei)`, therefore I have nothing of value and I'll return an empty iterator", which was inaccurate because it ignored the fact that there was a `d->e` prefix-transform rule. The reason why the tests are passing right now is because almost all bounds are set at iter creation time in the tests we've written, so SetBounds doesn't get called.

If we're transforming keys in reverse at the interface, we'd need to transform them back forward before we apply VSST file bounds in `constrainMeta`, and then in reverse again when we pass them to the `singleLevelIterator`. That's 3 one-way transformations for each bound change, so I think it's best to just leave the current logic in `constrainBounds` which just does it once and correctly.

Reply:
I see.

I'm still not happy with this though.

This feels like an abstraction leaking. Badly.

Lamme ponder a bit but if I can't find a better answer than just passing through (wrong, but then later right) raw keys I'll at least add giant comment.

Reply:
added optional `SetBoundsMaterialized` API for an underlying iter to indicate it wants these weird bounds and a big comment explaining the very subtle reason why. WDYT?